lib/exprengine.cpp

```diff
@@ -2497,17 +2544,11 @@ static std::string execute(const Token *start, const Token *end, Data &data)
             const Token *cond = tok->next()->astOperand2(); // TODO: C++17 condition
             const ExprEngine::ValuePtr condValue = executeExpression(cond, data);
 
-            bool alwaysFalse = false;
-            bool alwaysTrue = false;
-            if (auto b = std::dynamic_pointer_cast<ExprEngine::BinOpResult>(condValue)) {
-                alwaysFalse = !b->isTrue(&data);
-                alwaysTrue = !alwaysFalse && !b->isEqual(&data, 0);
-            }
-
             Data &thenData(data);
             Data elseData(data);
-            thenData.addConstraint(condValue, true);
-            elseData.addConstraint(condValue, false);
+            bool alwaysFalse = conditionAlwaysFalse(condValue, &data);
+            bool alwaysTrue = conditionAlwaysTrue(condValue, &data);
```

Comment:
I suggest that you short circuit here. If `alwaysFalse` is true then there is no need to ask z3..